orca/orca-spec.md

```diff
@@ -0,0 +1,278 @@
+considering the matrix of platforms, languages and virtualization/container environments. CPU and
+memory are, however, application agnostic and generally useful.
+
+Applications may arbitrarily define additional metrics as keys and units for their corresponding
+values.
+
+ORCA reports are ultimately aggregated into ClusterStats messages in LRS reports from the L7 load
+balancer to the corresponding management server.
+
+Without prescribing wire format, the following protobuf schema provides the data model for ORCA
+reports:
+
+See https://github.com/cncf/xds/blob/main/xds/data/orca/v3/orca_load_report.proto for the source
+of truth.
+
+```
+message OrcaLoadReport {
+  // CPU utilization expressed as a fraction of available CPU resources. This
+  // should be derived from the latest sample or measurement. The value may be
+  // larger than 1.0 when the usage exceeds the reporter dependent notion of
+  // soft limits.
+  double cpu_utilization = 1 [(validate.rules).double.gte = 0];
+
+  // Memory utilization expressed as a fraction of available memory
+  // resources. This should be derived from the latest sample or measurement.
+  double mem_utilization = 2 [(validate.rules).double.gte = 0, (validate.rules).double.lte = 1];
+
+  // Total RPS being served by an endpoint. This should cover all services that an endpoint is
+  // responsible for.
+  // Deprecated -- use ``rps_fractional`` field instead.
+  uint64 rps = 3 [deprecated = true];
+
+  // Application specific requests costs. Each value is an absolute cost (e.g. 3487 bytes of
+  // storage) associated with the request.
+  map<string, double> request_cost = 4;
+
+  // Resource utilization values. Each value is expressed as a fraction of total resources
+  // available, derived from the latest sample or measurement.
+  map<string, double> utilization = 5
+      [(validate.rules).map.values.double.gte = 0, (validate.rules).map.values.double.lte = 1];
+
+  // Total RPS being served by an endpoint. This should cover all services that an endpoint is
+  // responsible for.
+  double rps_fractional = 6 [(validate.rules).double.gte = 0];
+
+  // Total EPS (errors/second) being served by an endpoint. This should cover
+  // all services that an endpoint is responsible for.
+  double eps = 7 [(validate.rules).double.gte = 0];
+
+  // Application specific opaque metrics.
+  map<string, double> named_metrics = 8;
+
+  // Application specific utilization expressed as a fraction of available
+  // resources. For example, an application may report the max of CPU and memory
+  // utilization for better load balancing if it is both CPU and memory bound.
+  // This should be derived from the latest sample or measurement.
+  // The value may be larger than 1.0 when the usage exceeds the reporter
+  // dependent notion of soft limits.
+  double application_utilization = 9 [(validate.rules).double.gte = 0];
+}
+```
+
+# Reporting
+
+## Inline Per-Request Format
```

Comment:
I think we need to provide some guidance about when it's useful to choose per-request vs. OOB reporting.

I think we should recommend that the vast majority of applications use per-request reporting.  OOB reporting should be considered only for traffic patterns that do not result in per-request reports being delivered often enough to stay fresh, such as a client with very sparse traffic or an application whose traffic is mostly long-running streams.

We should also point out that OOB reporting is good for utilization but doesn't make much sense for query cost reporting.